include/swift/SIL/TypeLowering.h

```diff
@@ -965,10 +965,27 @@ class TypeConverter {
 #endif
 
   enum class ABIDifference : uint8_t {
-    // No ABI differences, function can be trivially bitcast to result type.
-    Trivial,
+    // Types have compatible calling conventions and representations, so can
+    // be trivially bitcast.
+    CompatibleRepresentation,
+    
+    // No convention differences, function can be cast via `convert_function`
+    // without a thunk.
+    //
+    // There may still be a representation difference between values of the
+    // compared function types. This means that, if two function types
+    // have a matching argument or return of function type with
+    // `SameCallingConvention`, then the outer function types may not themselves
+    // have the `SameCallingConvention` because they need a thunk to convert
+    // the inner function value representation.
+    CompatibleCallingConvention,
```

Comment:
The contrast between this and the cases below isn't clear.  Maybe something like:

> The types have a representation difference that can be resolved with `convert_function`, such as a possible difference in ptrauth discriminators.

We don't need to be coy about using ptrauth as an example.

The comment about nested positions applies to all these cases except `CompatibleRepresentation`, so maybe it should just go there.

Also you keep using `SameCallingConvention` in these comments, but it looks like you renamed that to `CompatibleCallingConvention`.

Longer-term question: should we just roll `thin_to_thick_function` into `convert_function`, so that we have one instruction for doing superficial conversions?

Reply:
Thanks, I'll clean up the comment in a follow up commit.